File: linked-lists/Node.js
class Node {
  constructor(d, n=null) {
    this._data = d;
    this._next = n;
  }

  get data() {
    return this._data;
  }

  set data(v) {
    this._data = v;
  }

  get next() {
    return this._next;
  }

  set next(n) {
    if ( (n === null) || ((n instanceof Node) && (n != this)) ) {
      this._next = n;
    } else {
      throw new Error('Next must be a Node and not the receiver');
    }
  }
}

module.exports = Node;

File: linked-lists/node.rb
class Node
  attr_reader :next, :value

  def initialize(value, n=nil)
    @value = value
    @next = n
  end

  def next=(n)
    if n.instance_of?(Node) && (n != self)
      @next = n
    else
      raise(ArgumentError, "Next must be a Node, and cannot be self")
    end
  end

  def value=(val)
    @value = val
  end
end
